### PR TITLE
Switch from autotools to cmake

### PR DESCRIPTION
**Pull request by stopiccot**
CMake benefits:
 - Much more cleaner and more readable syntax
 - Can generate Xcode and Visual Studio projects for OS X and Windows platforms respectively
 - Less files

What was reimplemented:
 - Configuration. All header and function checks. `unshield_config.h` generation.
 - pkg-config support.
 - Installing. Including versioned shared library symlink.
 - Travis CI integration. Build matrix for all 4 build configurations.

Anyway it's only an initial commit so feel free to test it and share any thoughts. Also I'm not sure that we need rebuild.sh and valgrind.sh files. Maybe we can just add those command line snippets to readme. But I left them for now.

### CODE REVIEW

**Comment by twogood** [comment]
I'm honored that you have taken time to do this, thank you!

I just need to adjust my mind a bit to this change...

**Comment by stopiccot** [comment]
That's ok feel free to play with it. Maybe you find something that I've missed. Btw to build project with cmake you need to run:

     cmake . && make

File: lib/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+add_library(libunshield ${LIBUNSHIELD_LIBRARY_TYPE} ${LIBUNSHIELD_HEADES} ${LIBUNSHIELD_SOURCES})
+
+if(USE_OUR_OWN_MD5)
+    target_link_libraries(libunshield ${ZLIB_LIBRARY} md5 convert_utf)
+else()
+    target_link_libraries(libunshield ${ZLIB_LIBRARY} ${OPENSSL_LIBRARIES} convert_utf)
+endif()
+
+set_target_properties(libunshield PROPERTIES OUTPUT_NAME unshiled)
```
Comment (twogood) [inline lib/CMakeLists.txt]: "unshield" maybe? :-)

File: rebuild.sh [hunk 2]
```diff
@@ -1,6 +1,4 @@
 #!/bin/sh
 set -x
 export CFLAGS="-Wall -Werror -ggdb3"
-./bootstrap &&
-./configure --prefix=/var/tmp/unshield --with-ssl &&
-make install
+cmake -DCMAKE_INSTALL_PREFIX:PATH=/var/tmp/unshield . && make && make install
```
Comment (twogood) [inline rebuild.sh]: :+1:

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+add_executable(unshiled "unshield.c")
+target_link_libraries(unshiled libunshield)
```
Comment (twogood) [inline src/CMakeLists.txt]: "unshield" :-)

**Comment by stopiccot** [comment]
Hi, any updates on this?

**Comment by twogood** [comment]
Well, did you see my comments to various parts of the pull request? :-)

**Comment by stopiccot** [comment]
Yes, but I thought that you found "unshield" :)

**Comment by stopiccot** [comment]
I thats seems to be the only question

**Comment by twogood** [comment]
Yes, I need "unshield" spelled correctly everywhere but the pull request spells it "unshiled" in a couple of places.

**Comment by stopiccot** [comment]
Fixed!

**Comment by twogood** [comment]
Great, busy week now, thanks for your patience!

**Comment by wdlkmpx** [comment]
It would be a good idea to restore the autotools files, they are better suited for cross-compiling, the toolchain doesn't even need any special apps installed.

The current cmake stuff requires some adjustments, I identified a fews issues, but I don't feel like fixing them, I'd rather use autotools, because I'm cross compiling and producing static binaries,

Imagine static binaries for
- linux i686
- linux x86_64
- linux armv6l
- linux aarch64
- win32
- win64

**Comment by twogood** [comment]
@wdlkmpx I don't have any strong preference but I have believed that CMake would work better with the various Microsoft Windows toolchains?

**Comment by wdlkmpx** [comment]
I don't want to remove the CMake stuff, I'll restore the autotools
stuff and see what I can do with it.

CMake might work better with Windows stuff, I don't know.

This is a cli app written in C, autotools makes sense. It doesn't
makes much sense in python, java, C# projects.

Well, there's a cross compiler for win32 called mingw. And it's
possible to create windows binaries from Linux.

For example zlib doesn't include an autotools project... it makes it
harder for me to do things, but with miniz I don't need zlib.

It would take a few hours to be able to compile static binaries for
several Linux arches, that can used for testing. I'm yet to add mingw
to my scriptzz

**Comment by twogood** [comment]
Yeah cross-compiling would be nice. I also think that Windows binary is more important than the possibility to compile on Windows.

**Comment by wdlkmpx** [comment]
One of these days I'll be opening a huge PR that includes the md5 stuff, offering 4 static linux binaries for x86 and ARM 

..for people to test in all kinds of machines, assuming the project compiles with musl, otherwise no PR will happen ...

**Comment by wdlkmpx** [comment]
as a side effect of that PR... the AppVeyor cmake builds will also succeed I think

**Comment by twogood** [comment]
Yeah, maybe going all-in on static linking is a good idea! It sure seems to work for _go_ applications.

**Pull request merged**